test/compliance/client-credential-workflow_test.js

```diff
@@ -0,0 +1,142 @@
+//   as it's not desired to have an extra fake-user representing a server just to pass validation
+const userDoc = { id: 'machine2-123456789', name: 'machine2' };
+db.saveUser(userDoc);
+
+const oAuth2Server = new OAuth2Server({
+  model: {
+    ...createModel(db),
+    getUserFromClient: async function (_client) {
+      // in a machine2machine setup we might not have a dedicated "user"
+      // but we need to return a truthy response to
+      const client = db.findClient(_client.id, _client.secret);
+      return client && { ...userDoc };
+    }
+  }
+});
+
+const clientDoc = db.saveClient({
+  id: 'client-credential-test-client',
+  secret: 'client-credential-test-secret',
+  grants: ['client_credentials']
+});
+
+const enabledScope = 'read write';
+
+describe('ClientCredentials Workflow Compliance (4.4)', function () {
+  describe('Access Token Request (4.4.1)', function () {
+    /**
+     * 4.4.2.  Access Token Request
+     *
+     *    The client makes a request to the token endpoint by adding the
+     *    following parameters using the "application/x-www-form-urlencoded"
+     *    format per Appendix B with a character encoding of UTF-8 in the HTTP
+     *    request entity-body:
+     *
+     *    grant_type
+     *          REQUIRED.  Value MUST be set to "client_credentials".
+     *
+     *    scope
+     *          OPTIONAL.  The scope of the access request as described by
+     *          Section 3.3.
+     *
+     *    The client MUST authenticate with the authorization server as
+     *    described in Section 3.2.1.
+     */
+    it('authenticates the client with valid credentials', async function () {
+      const response = new Response();
+      const request = createRequest({
+        body: {
+          grant_type: 'client_credentials',
+          scope: enabledScope
+        },
+        headers: {
+          'authorization': 'Basic ' + Buffer.from(clientDoc.id + ':' + clientDoc.secret).toString('base64'),
+          'content-type': 'application/x-www-form-urlencoded'
+        },
+        method: 'POST',
+      });
+
+      const token = await oAuth2Server.token(request, response);
+
+      response.status.should.equal(200);
+      response.headers.should.deep.equal( { 'cache-control': 'no-store', pragma: 'no-cache' });
+      response.body.token_type.should.equal('Bearer');
+      response.body.access_token.should.equal(token.accessToken);
+      response.body.expires_in.should.be.a('number');
+      response.body.scope.should.equal(enabledScope);
+      ('refresh_token' in response.body).should.equal(false);
+
+      token.accessToken.should.be.a('string');
+      token.accessTokenExpiresAt.should.be.a('date');
+      ('refreshToken' in token).should.equal(false);
+      ('refreshTokenExpiresAt' in token).should.equal(false);
+      token.scope.should.equal(enabledScope);
+
+      db.accessTokens.has(token.accessToken).should.equal(true);
+      db.refreshTokens.has(token.refreshToken).should.equal(false);
```

Comment:
why should there be no refresh token?

Reply:
The tests only cover [client credential grant (section 4.4)](https://datatracker.ietf.org/doc/html/rfc6749.html#section-4.4) and there is explicitly stated:

```
[4.4.3](https://datatracker.ietf.org/doc/html/rfc6749.html#section-4.4.3).  Access Token Response

   If the access token request is valid and authorized, the
   authorization server issues an access token as described in
   [Section 5.1](https://datatracker.ietf.org/doc/html/rfc6749.html#section-5.1).  A refresh token SHOULD NOT be included.  If the request
   failed client authentication or is invalid, the authorization server
   returns an error response as described in [Section 5.2](https://datatracker.ietf.org/doc/html/rfc6749.html#section-5.2).

   An example successful response:

     HTTP/1.1 200 OK
     Content-Type: application/json;charset=UTF-8
     Cache-Control: no-store
     Pragma: no-cache

     {
       "access_token":"2YotnFZFEjr1zCsicMWpAA",
       "token_type":"example",
       "expires_in":3600,
       "example_parameter":"example_value"
     }
```

Reply:
I think this is, because in CCG there is usually no human involved and the server already stores the secret, ready to be accessed for getting a new access token and the server can also check the expiration before making a call so one could say this is up to your implementation to obtain a new access token before expiration